Add module mail_forward

yajo opened the pull request:
Move mail_forward module from [odoo-grupoesoc-addons](https://github.com/grupoesoc/odoo-grupoesoc-addons).

This module allows the user to forward a message to another object of the database and/or to any contact. See README for details.

Commit:
Move mail_forward module from [odoo-grupoesoc-addons][1]

[1]: https://github.com/grupoesoc/odoo-grupoesoc-addons

Commit:
Add template.

pedrobaeza commented:
@rafaelbn, you'll be interested in this one.

dreispt commented on mail_forward/wizard/mail_forward.py:
```diff
@@ -0,0 +1,120 @@
+    _name = "mail_forward.compose.message"
+    _inherits = {"mail.compose.message": "original_wizard_id"}
+
+    _models = [
```
Can we have this in a setup table instead? It would make it easier to customize.

yajo replied:
I'll try.

pedrobaeza replied:
Or maybe use res.request.link table or another one created for this purpose.

yajo replied:
What does it do? Actually I was just adding a boolean to `ir.model` called `mail_forward_target`. Seems the easiest way.

pedrobaeza replied:
That table is a list of models allowed to be referenced by default in reference fields.

dreispt replied:
I agree with Pedro - that was the solution used in the core for similar needs, such as [`res.request`](https://github.com/odoo/odoo/search?utf8=%E2%9C%93&q=res.request) for "Reference" type fields.

Commit:
Get referenceable modules from res.request.link.

This follows better the Odoo standard and makes it easier to configure
by the administrator.

yajo commented:
If you agree with the current implementation, I'll start with the test suite.

pedrobaeza commented:
I think it's better to use a new table instead adding a flag on existing one, because you can want to add a new forward target, but not having on referenceable models.

yajo commented:
But then I'd have to hard-code them again. I thought that was the motivation of the change...

pedrobaeza commented:
They're imported as data, and you can add any record you want via XML or by UI (if views are provided). It's not the same "hard-coding".

yajo commented:
But other modules will not add themselves to that table. The nice thing about `res.request.link` is that for example you install `sale` and it takes care of adding the correct models to that table.

That could not be done the way you say unless developers create `foo_model` and `foo_model_mail_forward` as auto-install module when `mail_forward` and `foo_model` are installed.

pedrobaeza commented:
But with your flag, you force to set that flag programatically or manually if you want to forward.

yajo commented:
No. [`default=True`](https://github.com/OCA/social/pull/12/files#diff-39017bdf8ca51ef1f2027ebe237a81ecR12).

pedrobaeza commented:
OK, @dreispt, what do you think?

rafaelbn commented:
Rebuliding for testing! @pedrobaeza @dreispt @Yajo any technical modification will be applied? Thanks

rafaelbn commented:
Tested in runbot, could you add subject and To when forwarding?
´´´
----------Forwarded message----------
From: Administrator
Date: Mon Oct 05 2015 19:39:26 GMT+0200
´´´

> To:
> Subject:

rafaelbn commented:
One doubt:

![05-10-2015 19-42-38](https://cloud.githubusercontent.com/assets/8736623/10288492/7c44ff3a-6b99-11e5-91d0-6909be200daf.png)

yajo commented:
> any technical modification will be applied?

I have to write some tests.

> Tested in runbot, could you and subject and To when forwarding?

I don't understand you, sorry.

> One doubt [...]

When you forward the e-mail, you need to select a target. It can be:
- One or more contacts.
- One database object.
- Both.

So when you select a contact, the database object is not required anymore, and vice versa.

Commit:
Restructure module following guidelines.

Commit:
Add tests.

yajo commented:
Tests written. I hope Coveralls is happy now.

dreispt commented on mail_forward/README.rst:
```diff
@@ -0,0 +1,101 @@
+
+  * If you want to *remove* a model from the list, it's usually better to just
+    *disable* its check box instead of deleting it, because that record might
+    be used by other modules.
```
This design looks perfect to me.

dreispt commented on mail_forward/models/compose_message.py:
```diff
@@ -0,0 +1,97 @@
+        return result
+
+    @api.model
+    def models(self):
```
Choose a better name here. Maybe `_get_model_selection`

yajo replied:
Done.

dreispt commented on mail_forward/views/compose_message.xml:
```diff
@@ -0,0 +1,64 @@
+        <data>
+            <!-- It's required to fill either `partner_ids` or
+                 `destination_object_id` -->
+            <xpath expr="//field[@name='partner_ids']" position="attributes">
```
You can avoid an xpath here, and use a simpler `<field name="partner_ids" position="attributes">`

yajo replied:
I did not avoid xpath because I prefer it. It's a standard, it's more concise, better documented and it always works.

dreispt replied:
It's unnecessarily complicated, and Odoo provides a convenient shortcut. Odoo wise, it's also standard, well documented and always works.

yajo replied:
It does not seem so complicated to me. XPath can be used in any case, while the other method only for some. That's why I prefer XPath: one tool to any case. Sometimes there are more than one way to do the right thing. I prefer mine this time. If you don't like it, feel free to PR after merging.

dreispt replied:
We disagree, but not a blocking issue either...

pedrobaeza replied:
I also avoid xpath whenever I can. Yes, you use the same tag (xpath), but the variety of forms that the xpath can take makes it even more difficult to handle. For example, novice users tend to switch these xpath expressions without knowing exactly the difference:

```
<xpath expr="field[@name='partner_ids']"
<xpath expr="/field[@name='partner_ids']"
<xpath expr="//field[@name='partner_ids']"
```

yajo replied:
I always love standards. [XPath is one, backed by W3C](http://www.w3.org/TR/xpath/).

I also love using the right tool for the job. Quoting from that link before:

>  The primary purpose of XPath is to address parts of an XML document.

Which means that probably any developer that has needed to select a XML element in any app (not only Odoo) has met already XPath, which makes them less newbies in Odoo. I did not need to read any docs when I first saw this tag in an Odoo view, its purpose is obvious, but I needed to do it when using the shortcuts you say.

Any Odoo programmer will have to learn XPath sooner or later, since QWeb inheritance works only with it.

But the most important thing is:

![hakuna-matata](https://cloud.githubusercontent.com/assets/973709/10453020/60d07ebc-71a9-11e5-8f47-b69d0f426dd9.png)

dreispt commented:
Congratulations, looks very good.

Commit:
Change model selection method name.

dreispt commented:
:+1:

rafaelbn commented:
Could you add this https://github.com/OCA/social/pull/12#issuecomment-145609508 ?

Commit:
Easier way to generate forward header.

Add CC and Subject if available.

Commit:
Use strict js.

Commit:
Update translations.

yajo commented:
Done.

hbrunn commented:
:+1:
